File: modules/angular2/src/core/linker/directive_lifecycle_reflector.ts
import {Type} from 'angular2/src/core/facade/lang';
import {LifecycleHooks} from './interfaces';

export function hasLifecycleHook(lcInterface: LifecycleHooks, token): boolean {
  if (!(token instanceof Type)) return false;

  var proto = (<any>token).prototype;

  switch (lcInterface) {
    case LifecycleHooks.AfterContentInit:
      return !!proto.afterContentInit;
    case LifecycleHooks.AfterContentChecked:
      return !!proto.afterContentChecked;
    case LifecycleHooks.AfterViewInit:
      return !!proto.afterViewInit;
    case LifecycleHooks.AfterViewChecked:
      return !!proto.afterViewChecked;
    case LifecycleHooks.OnChanges:
      return !!proto.onChanges;
    case LifecycleHooks.DoCheck:
      return !!proto.doCheck;
    case LifecycleHooks.OnDestroy:
      return !!proto.onDestroy;
    case LifecycleHooks.OnInit:
      return !!proto.onInit;
    default:
      return false;
  }
}

File: modules/angular2/src/compiler/directive_lifecycle_reflector.dart
library angular2.src.core.compiler.directive_lifecycle_reflector;

import 'package:angular2/src/core/reflection/reflection.dart';
import 'package:angular2/src/core/metadata/lifecycle_hooks.dart';

const INTERFACES = const {
  LifecycleHooks.OnInit: OnInit,
  LifecycleHooks.OnDestroy: OnDestroy,
  LifecycleHooks.DoCheck: DoCheck,
  LifecycleHooks.OnChanges: OnChanges,
  LifecycleHooks.AfterContentInit: AfterContentInit,
  LifecycleHooks.AfterContentChecked: AfterContentChecked,
  LifecycleHooks.AfterViewInit: AfterViewInit,
  LifecycleHooks.AfterViewChecked: AfterViewChecked,
};

bool hasLifecycleHook(LifecycleHooks interface, token) {
  if (token is! Type) return false;
  Type interfaceType = INTERFACES[interface];
  return reflector.interfaces(token).contains(interfaceType);
}
